server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -113,17 +113,24 @@ void sendRequest(
             isHandshake,
             compressionScheme
         );
-        if (request.tryIncRef() == false) {
-            assert false : "request [" + request + "] has been released already";
-            throw new AlreadyClosedException("request [" + request + "] has been released already");
-        }
-        sendMessage(channel, message, ResponseStatsConsumer.NONE, () -> {
-            try {
-                messageListener.onRequestSent(node, requestId, action, request, options);
-            } finally {
-                request.decRef();
+        final Releasable onAfter;
+        if (request instanceof BytesTransportRequest) {
```

Comment:
I'd rather we pushed this down into `OutboundMessage#serialize` that already has the special case for `BytesTransportRequest`, i.e. attaching this ref to the outbound buffers directly.